nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/configuration/JwtAuthenticationSecurityConfiguration.java

```diff
@@ -0,0 +1,195 @@
+        final StateManager stateManager = stateManagerProvider.getStateManager(StandardJwtRevocationService.class.getName());
+        return new StandardJwtRevocationService(stateManager);
+    }
+
+    @Bean
+    public JwtLogoutListener jwtLogoutListener() {
+        return new StandardJwtLogoutListener(jwtDecoder(), jwtRevocationService());
+    }
+
+    @Bean
+    public JWTProcessor<SecurityContext> jwtProcessor() {
+        final DefaultJWTProcessor<SecurityContext> jwtProcessor = new DefaultJWTProcessor<>();
+        jwtProcessor.setJWSKeySelector(jwsKeySelector());
+        jwtProcessor.setJWTClaimsSetVerifier(claimsSetVerifier());
+        return jwtProcessor;
+    }
+
+    @Bean
+    public JWSKeySelector<SecurityContext> jwsKeySelector() {
+        return new StandardJWSKeySelector<>(verificationKeySelector());
+    }
+
+    @Bean
+    public JWTClaimsSetVerifier<SecurityContext> claimsSetVerifier() {
+        return new DefaultJWTClaimsVerifier<>(null, REQUIRED_CLAIMS);
+    }
+
+    @Bean
+    public StandardJwtAuthenticationConverter jwtAuthenticationConverter() {
+        return new StandardJwtAuthenticationConverter(authorizer, idpUserGroupService, niFiProperties);
+    }
+
+    @Bean
+    public BearerTokenProvider bearerTokenProvider() {
+        return new StandardBearerTokenProvider(jwsSignerProvider());
+    }
+
+    @Bean
+    public StandardJwsSignerProvider jwsSignerProvider() {
+        return new StandardJwsSignerProvider(verificationKeySelector());
+    }
+
+    @Bean
+    public StandardVerificationKeySelector verificationKeySelector() {
+        return new StandardVerificationKeySelector(verificationKeyService(), niFiProperties.getSecurityUserJwsKeyRotationPeriod());
+    }
+
+    @Bean
+    public VerificationKeyService verificationKeyService() {
+        final StateManager stateManager = stateManagerProvider.getStateManager(StandardVerificationKeyService.class.getName());
+        return new StandardVerificationKeyService(stateManager);
+    }
+
+    @Bean
+    public KeyGenerationCommand keyGenerationCommand() {
+        final KeyGenerationCommand command = new KeyGenerationCommand(jwsSignerProvider(), verificationKeySelector());
+        commandScheduler().scheduleAtFixedRate(command, niFiProperties.getSecurityUserJwsKeyRotationPeriod());
+        return command;
+    }
+
+    @Bean
+    public KeyExpirationCommand keyExpirationCommand() {
+        final KeyExpirationCommand command = new KeyExpirationCommand(verificationKeyService());
+        commandScheduler().scheduleAtFixedRate(command, niFiProperties.getSecurityUserJwsKeyRotationPeriod());
+        return command;
+    }
+
+    @Bean
+    public RevocationExpirationCommand revocationExpirationCommand() {
+        final RevocationExpirationCommand command = new RevocationExpirationCommand(jwtRevocationService());
+        commandScheduler().scheduleAtFixedRate(command, niFiProperties.getSecurityUserJwsKeyRotationPeriod());
```

Comment:
In the unlikely scenario that the user configures the key rotation period to be longer than the login identity provider's authentication expiration, this wouldn't expire the revocation in time, right?  Perhaps that could be considered a user error, though.

Reply:
The Revocation Expiration Command cleans up the cache of revoked tokens to avoid keeping unnecessary references. The revoked tokens include the JSON Web Token identifier and associated expiration, so the storage used is minimal, and keeping references around for a longer than the expiration should not be a problem.